File: tools/upload-scripts/thumb-handler.php
<?php

$uri = $_SERVER['REQUEST_URI'];
	
# Is this a thumbnail?
if ( preg_match('!^(?:http://upload.wikimedia.org)?/([\w-]*)/([\w-]*)/thumb(/archive|)/\w/\w\w/([^/]*)/' . 
	'(page(\d*)-)*(\d*)px-([^/]*)$!', $uri, $matches ) )
{
	list( $all, $site, $lang, $arch, $filename, $pagefull, $pagenum, $size, $fn2 ) = $matches;
	$params = array( 
		'f' => $filename,
		'width' => $size
	);
	if ( $pagenum ) {
		$params['page'] = $pagenum;
	}
	if ( $arch ) {
		$params['archived'] = 1;
	}
} elseif ( preg_match('!^(?:http://upload.wikimedia.org)?/([\w-]*)/([\w-]*)/thumb(/archive|)/\w/\w\w/([^/]*\.(?:(?i)ogg|ogv|oga))/' . 
	'(mid|seek(?:=|%3D|%3d)\d+)-([^/]*)$!', $uri, $matches ) ) 
{
	list( $all, $site, $lang, $arch, $filename, $timeFull, $fn2 ) = $matches;
	$params = array( 'f' => $filename );
	if ( $timeFull != 'mid' ) {
		list( $seek, $thumbtime ) = explode( '=', urldecode( $timeFull ), 2 );
		$params['thumbtime'] = $thumbtime;
	}
	if ( $arch ) {
		$params['archived'] = 1;
	}
} else {
	# No, display standard 404
	header( 'X-Debug: no regex match' );
	require_once( '404.php' );
	exit;
}

# Do some basic checks on the filename
if ( preg_match( '/[\x80-\xff]/', $uri ) ) {
	header( 'HTTP/1.0 400 Bad request' );
	header( 'Content-Type: text/html' );
	echo "<html><head><title>Bad request</title></head><body>" . 
		"The URI contained bytes with the high bit set, this is not allowed." . 
		"</body></html>";
	exit;
}
if ( strpos( $filename, '%20' ) !== false ) {
	header( 'HTTP/1.0 404 Not found' );
	header( 'Content-Type: text/html' );
	header( 'X-Debug: filename contains a space' );
	echo "<html><head><title>Not found</title></head><body>" . 
		"The URL contained spaces, we don't have any thumbnail files with spaces." . 
		"</body></html>";
	exit;
}

/* Breaks SVG and ogg
if ( $filename != $fn2 ) {
	require_once( '404.php' );
	exit;
}
*/

# Determine hostname
if ( $site == 'wikipedia' ) {
	switch ( $lang ) {
		case 'meta':
		case 'commons':
		case 'internal':
		case 'grants':
		case 'wikimania2006':
			$host = "$lang.wikimedia.org";
			break;
		case 'mediawiki':
			$host = "www.mediawiki.org";
			break;
		case 'sources':
			$host = 'wikisource.org';
			break;
		default:
			$host = "$lang.wikipedia.org";


	}
} else {
	$sitelang = str_replace( '-', '.', $lang );
	$host = "$sitelang.$site.org";
}

# Send request using CURL
$reqURL = "http://$host/w/thumb.php?";
$first = true;
foreach ( $params as $name => $value ) {
	if ( $first ) {
		$first = false;
	} else {
		$reqURL .= '&';
	}
	// Note: value is already urlencoded
	$reqURL .= "$name=$value";
}

header( "X-Wikimedia-Thumb: $reqURL" );


$ch = curl_init( $reqURL );
curl_setopt($ch, CURLOPT_PROXY, '10.2.1.21:80');

# Set an XFF header for abuse tracking
# Use $_SERVER not apache_request_headers beccause this runs under fastcgi
if ( !empty( $_SERVER['HTTP_X_FORWARDED_FOR'] ) ) {
	$xff = $_SERVER['HTTP_X_FORWARDED_FOR'] . ', ' . $_SERVER['REMOTE_ADDR'];
} else {
	$xff = $_SERVER['REMOTE_ADDR'];
}

$headers = array(
	"X-Forwarded-For: " . str_replace( "\n", '', $xff ),
	"X-Original-URI: " . str_replace( "\n", '', $uri )
);

# Pass through some other headers
$passthrough = array( 
	'If-Modified-Since',
	'Referer',
	'User-Agent'
);
foreach ( $passthrough as $headerName ) {
	$serverVarName = 'HTTP_' . str_replace( '-', '_', strtoupper( $headerName ) );
	if ( !empty( $_SERVER[$serverVarName] ) ) {
		$headers[] = $headerName . ': ' . 
			str_replace( "\n", '', $_SERVER[$serverVarName] );
	}
}

curl_setopt( $ch, CURLOPT_HTTPHEADER, $headers );
curl_setopt( $ch, CURLOPT_RETURNTRANSFER, true );

# Leave it long enough to generate a ulimit timeout in ordinary cases
# But short enough to avoid a local PHP timeout (55s)
curl_setopt( $ch, CURLOPT_TIMEOUT, 53 );

$text = curl_exec( $ch );

# Send it on to the client
$errno = curl_errno( $ch );
$contentType = curl_getinfo( $ch, CURLINFO_CONTENT_TYPE );
$httpCode = curl_getinfo( $ch, CURLINFO_HTTP_CODE );
if ( $errno ) {
	header( 'HTTP/1.1 500 Internal server error' );
	header( 'Cache-Control: no-cache' );
	$contentType = 'text/html';
	$error = htmlspecialchars( curl_error( $ch ) );
	$text = <<<EOT
<html>
<head><title>Thumbnail error</title></head>
<body>Error retrieving thumbnail from scaling server: $error</body>
</html>
EOT;
} elseif ( $httpCode == 304 ) {
	header( 'HTTP/1.1 304 Not modified' );
	$contentType = '';
	$text = '';
} elseif ( strval( $text ) == '' ) {
	header( 'HTTP/1.1 500 Internal server error' );
	header( 'Cache-Control: no-cache' );
	$contentType = 'text/html';
	$error = htmlspecialchars( curl_error( $ch ) );
	$text = <<<EOT
<html>
<head><title>Thumbnail error</title></head>
<body>Error retrieving thumbnail from scaling server: empty response</body>
</html>
EOT;
} elseif ( $httpCode == 404 ) {
	header( 'HTTP/1.1 404 Not found' );
	// header( 'Cache-Control: no-cache' );
	header( 'Cache-Control: s-maxage=300, must-revalidate, max-age=0' );
} elseif ( $httpCode != 200 || substr( $contentType, 0, 9 ) == 'text/html' || substr( $text, 0, 5 ) == '<html' ) {
	# Error message, suppress cache
	header( 'HTTP/1.1 500 Internal server error' );
	header( 'Cache-Control: no-cache' );
} else {
	# Cache the file locally if this is a cache server
	$uname = posix_uname();
	$server = $uname['nodename'];
	if ( $server != 'amane.pmtpa.wmnet' && $server != 'storage1' ) {
		$dest = pathFromUrl( $_SERVER['REQUEST_URI'] );
		if( $dest ) {
			if ( strpos( $dest, '..' ) === false ) {
				# Make directory and parents
				for ( $i = 0, $path = dirname( $dest ); !file_exists( $path ) && $i < 6; $path = dirname( $path ), $i++ ) {
					@mkdir( $path );
				}
				@file_put_contents( $dest, $text );
			}
		} else {
			
			header( "X-Debug: got invalid input path" );
		}
	}
}

if ( !$contentType ) {
	header( 'Content-Type:' );
} else {
	header( "Content-Type: $contentType" );
}
print $text;
curl_close( $ch );

function pathFromUrl( $url ) {
	if( preg_match( '!^(?:http://upload.wikimedia.org)?/([\w-]*)/([\w-]*)/thumb(/archive|)/(\w)/(\w\w)/([^/]*)/([^/]*)$!',
		$url, $matches ) ) {
		$parts = array_map( 'rawurldecode', $matches );
		list( $all, $site, $lang, $arch, $hash1, $hash2, $filename, $fn2 ) = $parts;
				
		$md5 = md5( $filename );
		if( $hash1 != substr( $md5, 0, 1 ) ) return false;
		if( $hash2 != substr( $md5, 0, 2 ) ) return false;

		$testNames = array( $filename, "$filename.png", "$filename.jpg" );
		$good = false;
		foreach ( $testNames as $testName ) {
			if ( substr( $fn2, -strlen( $testName ) ) == $testName ) {
				$good = true;
				break;
			}
		}
		
		if( $good ) {
			$thumbPath = array( '', 'export', 'upload', $site, $lang, 'thumb', $arch, $hash1, $hash2, $filename, $fn2 );
			if ( !checkPathComponents( $thumbPath ) ) return false;
			return implode( '/', $thumbPath );
		}
	}
	return false;
}

function checkPathComponents( $components ) {
	foreach( $components as $component ) {
		if( strpos( $component, '/' ) !== false ) return false;
		if( $component == '.' ) return false;
		if( $component == '..' ) return false;
	}
	return true;
}

?>

File: tools/upload-scripts/nsapi/thumb-handler.c
/* Copyright (c) 2008 River Tarnell <river@loreley.flyingparchment.org.uk>. */
/*
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely. This software is provided 'as-is', without any express or implied
 * warranty.
 */
/* $Id$ */

/*
 * This is a C implementation of the thumb-handler.php script as an NSAPI
 * module.  Running as NSAPI is faster, and more efficient (as we don't need
 * to keep a load of PHP processes around).
 *
 * To use it, first load the module in magnus.conf:
 *   Init fn="load-modules" shlib="thumb-handler.so" funcs="thumb-handler,thumb-handler-init"
 *   Init fn="thumb-handler-init" 
 * Then invoke it in obj.conf:
 *   Service method="(GET|HEAD)" type="*~magnus-internal/*" fn="thumb-handler"
 */

#include	<netdb.h>
#include	<nsapi.h>
#include	<stdio.h>
#include	<unistd.h>
#include	<string.h>
#include	<pcre.h>
#include	<curl/curl.h>
#include	<glib.h>

/* Size of the output array for pcre_exec */
#define OVECSIZE	15

struct call_context {
	Session	*sn;
	Request	*rq;
	char *data;
	size_t sz;
};

static char *strip_hostname(char *uri);
static int re_match(Session *, Request *, char const *re, char const *subj, char ***output);
static int prepare_curl(Session *, Request *, CURL *, GString *requrl, struct curl_slist **);
static int add_headers(Session *, Request *rq, CURL *curl, struct curl_slist **);
static char *make_xff_header(Session *, Request *rq);
static GString *pblock_to_query(pblock const *);
static GString *hostname_from_site_lang(char const *site, char const *lang);
static GString *thumb_url_from_path(Session *sn, Request *rq, char const *path);
static size_t write_data(void *ptr, size_t, size_t, void *);
static void send_error(Session *sn, Request *rq, int code, char const *fmt, ...);

#ifdef __GNUC__
# define UNUSED __attribute__((unused))
#else
# define UNUSED
#endif

/*
 * cURL callback that writes data to the client
 */
static size_t
write_data(ptr, size, num, data)
	void *ptr, *data;
	size_t size, num;
{
size_t	sz = size * num;
struct call_context *ctx = (struct call_context *)data;
	ctx->data = realloc(ctx->data, ctx->sz + sz);
	if (ctx->data == NULL) {
		log_error(LOG_FAILURE, "thumb-handler", ctx->sn, ctx->rq, "out of memory");
		return 0;
	}
	memcpy(ctx->data + ctx->sz, ptr, sz);
	ctx->sz += sz;
	return sz;
}

/*
 * Handle initialisation on server startup.
 */
int thumb_handler_init(pblock *pb UNUSED, Session *sn, Request *rq)
{
	if (curl_global_init(CURL_GLOBAL_NOTHING) != 0)
		return REQ_ABORTED;
	log_error(LOG_INFORM, "thumb-handler", sn, rq, "thumb-handler $Id$ ready");
	return REQ_PROCEED;
}

static void
send_error(
	Session *sn,
	Request *rq,
	int code,
	char const *fmt,
	...)
{
va_list	 ap;
GString	*err = g_string_new(NULL);
	g_string_assign(err,
"<html>\n"
"<head><title>Thumbnail error</title></head>\n"
"<body>\n");
	
	va_start(ap, fmt);
	g_string_append_vprintf(err, fmt, ap);
	va_end(ap);

	g_string_append(err, "</body>\n</html>\n");
	pblock_nvinsert("content-type", "text/html", rq->srvhdrs);
	pblock_nvinsert("cache-control", "no-cache", rq->srvhdrs);
	protocol_status(sn, rq, code, NULL);
	if (protocol_start_response(sn, rq) != REQ_NOACTION)
		net_write(sn->csd, err->str, err->len);
	g_string_free(err, TRUE);
}

/*
 * This function handles the actual request.
 */
int
thumb_handler(pblock *pb UNUSED, Session *sn, Request *rq)
{
char			*uri, *uripath;
GString			*thumburl;
CURL			*curl;
struct call_context	 context;
char			 curlerror[CURL_ERROR_SIZE];
long			 httpcode;
char			*contenttype;
struct curl_slist	*headers = NULL;

	/* Get the URI from the request */
	if ((uri = pblock_findval("uri", rq->reqpb)) == NULL) {
		/* Shouldn't happen, but just in case... */
		log_error(LOG_WARN, "thumb-handler", sn, rq,
			   "received a request with a NULL URI");
		return REQ_ABORTED;
	}

	if (util_uri_is_evil(uri)) {
		log_error(LOG_WARN, "thumb-handler", sn, rq,
			   "received a request with an evil URI: %s",
			   uri);
		return REQ_ABORTED;
	}
	
	/*
	 * If the URI starts with a hostname, i.e. http://example.com/...,
	 * strip it off.  Squid sends requests like this.
	 */

	if ((uripath = strip_hostname(uri)) == NULL) {
		/*
		 * A request like "GET http://example.com" will leave uripath empty.
		 */
		log_error(LOG_WARN, "thumb-handler", sn, rq,
			   "received a request with invalid URI: %s",
			   uri);
		return REQ_ABORTED;
	}
	
	if ((thumburl = thumb_url_from_path(sn, rq, uripath)) == NULL)
		return REQ_ABORTED;
	
	pblock_nvinsert("x-wikimedia-thumb", thumburl->str, rq->srvhdrs);
	
	/*
	 * Fetch the image.
	 */
	curl = curl_easy_init();
	if (!curl) {
		log_error(LOG_WARN, "thumb-handler", sn, rq,
			   "failed to initialise libcURL");
		send_error(sn, rq, PROTOCOL_SERVER_ERROR, "Error initialising cURL");
		curl_easy_cleanup(curl);
		g_string_free(thumburl, TRUE);
		return REQ_PROCEED;
	}

	if (prepare_curl(sn, rq, curl, thumburl, &headers) == -1) {
		send_error(sn, rq, PROTOCOL_SERVER_ERROR, "Error preparing cURL");
		curl_easy_cleanup(curl);
		g_string_free(thumburl, TRUE);
		return REQ_PROCEED;
	}
	
	memset(&context, 0, sizeof(context));
	context.sn = sn;
	context.rq = rq;
	
	curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_data);
	curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)&context);
	curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, curlerror);
	
	if (curl_easy_perform(curl) != 0) {
		/*
		 * Some kind of cURL error occurred.
		 */
		send_error(sn, rq, PROTOCOL_SERVER_ERROR, 
			    "Error retrieving thumbnail from scaling servers: %s",
			    curlerror);
		
		log_error(LOG_WARN, "thumb-handler", sn, rq,
			   "cURL failed to fetch image [url=%s]: %s",
			   thumburl->str, curlerror);
		free(context.data);
		curl_easy_cleanup(curl);
		curl_slist_free_all(headers);
		g_string_free(thumburl, TRUE);
		return REQ_PROCEED;
	}

	g_string_free(thumburl, TRUE);

	curl_easy_getinfo(curl, CURLINFO_CONTENT_TYPE, &contenttype);
	curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &httpcode);

	curl_slist_free_all(headers);
	
	switch (httpcode) {
	case 304:
		protocol_status(sn, rq, PROTOCOL_NOT_MODIFIED, NULL);
		curl_easy_cleanup(curl);
		free(context.data);
		return REQ_PROCEED;
	
	case 404:
		log_error(LOG_WARN, "thumb-handler", sn, rq,
			   "Response code was 404");
		send_error(sn, rq, PROTOCOL_NOT_FOUND, 
			   "Received 404 response from scaling server");
		curl_easy_cleanup(curl);
		free(context.data);
		return REQ_PROCEED;
	}
	
	if (context.sz == 0) {
		log_error(LOG_WARN, "thumb-handler", sn, rq,
			  "Empty response from scaling server");
		send_error(sn, rq, PROTOCOL_SERVER_ERROR, 
		    "Error retrieving thumbnail from scaling servers: empty response");
		curl_easy_cleanup(curl);
		free(context.data);
		return REQ_PROCEED;
	}
	
	if (httpcode != 200 || strncmp(contenttype, "text/html", 9) == 0 ||
	    (context.sz > 5 && memcmp(context.data, "<html", 5) == 0)) {
		protocol_status(sn, rq, PROTOCOL_SERVER_ERROR, NULL);
		pblock_nvinsert("cache-control", "nocache", rq->srvhdrs);
	} else {
		protocol_status(sn, rq, PROTOCOL_OK, NULL);
	}
	
	pblock_nvinsert("content-type", contenttype, rq->srvhdrs);
	if (protocol_start_response(sn, rq) != REQ_NOACTION)
		net_write(sn->csd, context.data, context.sz);
	curl_easy_cleanup(curl);
	free(context.data);
	
	return REQ_PROCEED;
}

static int
prepare_curl(sn, rq, curl, requrl, headers)
	Session *sn;
	Request *rq;
	CURL *curl;
	GString *requrl;
	struct curl_slist **headers;
{
	curl_easy_setopt(curl, CURLOPT_NOSIGNAL, (long) 1);
	curl_easy_setopt(curl, CURLOPT_TIMEOUT, (long) 53);
	curl_easy_setopt(curl, CURLOPT_URL, requrl->str);
	curl_easy_setopt(curl, CURLOPT_PROXY, "10.2.1.21:80");

	if (add_headers(sn, rq, curl, headers) == -1)
		return -1;
	
	return 0;
}

static int
add_headers(sn, rq, curl, headers)
	Session *sn;
	Request *rq;
	CURL *curl;
	struct curl_slist **headers;
{
GString			*str;
const char		**h;
char			*xff;
static char const *passthrough[] = {
	"if-modified-since",
	"referer",
	"user-agent",
	NULL
};
	if ((xff = make_xff_header(sn, rq)) == NULL)
		return -1;
	
	*headers = curl_slist_append(*headers, xff);
	
	str = g_string_new(NULL);
	g_string_printf(str, "X-Original-URI: %s", pblock_findval("uri", rq->reqpb));
	*headers = curl_slist_append(*headers, str->str);
	g_string_free(str, TRUE);
	
	for (h = passthrough; *h; ++h) {
	char	*s;
		if ((s = pblock_findval(*h, rq->headers)) != NULL) {
		GString	*hdr;
			hdr = g_string_new(NULL);
			g_string_printf(hdr, "%s: %s", *h, s);
			*headers = curl_slist_append(*headers, hdr->str);
			g_string_free(hdr, TRUE);
		}
	}
	
	curl_easy_setopt(curl, CURLOPT_HTTPHEADER, *headers);
	return 0;
}

static char *
make_xff_header(sn, rq)
	Session *sn;
	Request *rq;
{
	/*
	 * Set an XFF header for abuse tracking.
	 */
char	*xff, *s, *r;
	if ((s = pblock_findval("x-forwarded-for", rq->headers)) != NULL) {
		char *raddr = pblock_findval("ip", sn->client);
		if (raddr == NULL) {
			log_error(LOG_FAILURE, "thumb-handler", sn, rq,
				   "could not find remote address for client");
			return NULL;
		}
		
		char *xff = CALLOC(strlen(raddr) + strlen(s) + 3);
		if (xff == NULL) {
			log_error(LOG_FAILURE, "thumb-handler", sn, rq, "out of memory");
			return NULL;
		}
		sprintf(xff, "%s, %s", s, raddr);
	} else {
		xff = pblock_findval("ip", sn->client);
		if (xff == NULL) {
			log_error(LOG_FAILURE, "thumb-handler", sn, rq,
				   "could not find remote address for client");
			return NULL;
		}
	}
	
	r = CALLOC(sizeof("X-Forwarded-For: ") + strlen(xff));
	if (r == NULL) {
		log_error(LOG_FAILURE, "thumb-handler", sn, rq, "out of memory");
		return NULL;
	}
	sprintf(r, "X-Forwarded-For: %s", xff);
	
	return r;
}

/*
 * Given a string like "http://example.com/a/b/c", remove
 * "http://example.com" and leave "/a/b/c".
 */
static char *
strip_hostname(uri)
	char *uri;
{
	if (strncmp(uri, "http://", 7) == 0)
		return strchr(uri + 7, '/');
	else
		return uri;
}

static int 
re_match(sn, rq, rex, subj, matches)
	Session *sn;
	Request *rq;
	char const *rex, *subj;
	char ***matches;
{
pcre		*re;
int		 rc, i;
int		 ovector[OVECSIZE * 3];
char const	*error;
int		 erroffset;

	re = pcre_compile(rex, 0, &error, &erroffset, NULL);
	
	if (re == NULL) {
		log_error(LOG_WARN, "thumb-handler", sn, rq,
			   "pcre_compile failed (offset=%d): %s",
			   erroffset, error);
		return -1;
	}
	
	rc = pcre_exec(re, NULL, subj, strlen(subj), 0, 0,
			ovector, OVECSIZE * 3);
	
	pcre_free(re);
	
	if (rc <= 0) {
		switch (rc) {
		case PCRE_ERROR_NOMATCH:
			return -1;
			
		default:
			log_error(LOG_WARN, "thumb-handler", sn, rq,
				   "pcre_exec failed: %d", rc);
			return rc;
		}
	}
	
	*matches = CALLOC(sizeof(char *) * rc);
	if (*matches == NULL) {
		log_error(LOG_FAILURE, "thumb-handler", sn, rq, "out of memory");
		return -1;
	}

	for (i = 0; i < rc; ++i) {
		char const *start = subj + ovector[2 * i];
		int len = ovector[2*i + 1] - ovector[2 * i];
		(*matches)[i] = CALLOC(len + 1);
		if ((*matches)[i] == NULL) {
			log_error(LOG_FAILURE, "thumb-handler", sn, rq, "out of memory");
			return -1;
		}

		memcpy((*matches)[i], start, len);
	}
	
	return rc;
}
	
static GString *
thumb_url_from_path(sn, rq, path)
	Session *sn;
	Request *rq;
	char const *path;
{
char		*site, *lang, *width, *filename, *page;
pblock		*params;
char		**matches;
GString		*requrl;
GString		*host, *paramstr;
	/*
	 * Extract the various parts from the URI.  These regexes come from
	 * thumb-handler.php.
	 */

#define THUMB_REGEXP \
	"^/([\\w-]*)/([\\w-]*)/thumb(/archive|)/\\w/\\w\\w/([^/]*)/(page(\\d*)-)*(\\d*)px-([^/]*)$"

	/*
	 * Use a pblock to store the various parameters we need to send
	 * to thumb.php.
	 */
	params = pblock_create(19);

	/* First, look for a thumbnail */
	if (re_match(sn, rq, THUMB_REGEXP, path, &matches) > 0) {
		site = matches[1];
		lang = matches[2];
		filename = matches[4];
		page = matches[6];
		width = matches[7];
	
		pblock_nvinsert("f", filename, params);
		pblock_nvinsert("width", width, params);
		if (*page)
		pblock_nvinsert("page", page, params);
		if (matches[3][0])
			pblock_nvinsert("archived", "1", params);
	
#define MEDIA_REGEXP							\
	"^/([\\w-]*)/([\\w-]*)/thumb(/archive|)/\\w/\\w\\w/"		\
	"([^/]*\\.(?:(?i)ogg|ogv|oga))/(mid|seek(?:=|%3D|%3d)(\\d+))-([^/]*)$"
	} else if (re_match(sn, rq, MEDIA_REGEXP, path, &matches) > 0) {
		site = matches[1];
		lang = matches[2];
		filename = matches[4];
		pblock_nvinsert("f", filename, params);
		if (strcmp(matches[5], "mid"))
			pblock_nvinsert("thumbtime", matches[6], params);
		if (matches[3][0])
			pblock_nvinsert("archived", "1", params);
	} else {
		pblock_nvinsert("x-debug", "No regex match", rq->srvhdrs);
		send_error(sn, rq, PROTOCOL_NOT_FOUND, "No regex match for filename");
		return NULL;
	}

	if (strstr(filename, "%20") != NULL) {
		pblock_nvinsert("x-debug", "filename contains a space", rq->srvhdrs);
		send_error(sn, rq, PROTOCOL_NOT_FOUND, "Filename contains a space");
		return NULL;
	}

	/*
	 * Build the URL from the params pblock.
	 */
	host = hostname_from_site_lang(site, lang);
	paramstr = pblock_to_query(params);

	requrl = g_string_new(NULL);
	g_string_sprintf(requrl, "http://%s/thumb.php?%s", host->str, paramstr->str);
	
	g_string_free(host, TRUE);
	g_string_free(paramstr, TRUE);

	return requrl;
}

static GString *
hostname_from_site_lang(site, lang)
	char const *site, *lang;
{
GString	*host;

	host = g_string_new(NULL);
	if (strcmp(site, "wikipedia") == 0) {
		if (strcmp(lang, "meta") == 0 ||
		    strcmp(lang, "commons") == 0 ||
		    strcmp(lang, "internal") == 0 ||
		    strcmp(lang, "grants") == 0 ||
		    strcmp(lang, "wikimania2006") == 0)
		{
			g_string_printf(host, "%s.wikimedia.org", lang);
		} else if (strcmp(lang, "mediawiki") == 0) {
			g_string_assign(host, "www.mediawiki.org");
		} else if (strcmp(lang, "sources") == 0) {
			g_string_assign(host, "wikisource.org");
		} else {
			g_string_printf(host, "%s.wikipedia.org", lang);
		}
	} else {
		char *l = STRDUP(lang);
		char *p = l;
		while ((p = strchr(p, '-')) != NULL)
			*p = '.';
		g_string_printf(host, "%s.%s.org", l, site);
	}
	
	return host;
}

static GString *
pblock_to_query(pb)
	pblock const *pb;
{
GString	*query;
int	 i, first = 1;
	query = g_string_new(NULL);
	
	/* Iterate the pblock hash table */
	for (i = 0; i < pb->hsize; ++i) {
		struct pb_entry *entry;

		for (entry = pb->ht[i]; entry; entry = entry->next) {
			if (!first)
				g_string_append(query, "&");
			else
				first = 0;
			
			g_string_append_printf(query, "%s=%s",
				entry->param->name, entry->param->value);
		}
	}

	return query;
}